### PR TITLE
[`core`] Fix `use_reentrant` issues

### PR DESCRIPTION
**Pull request by younesbelkada**
Partially fixes: https://github.com/huggingface/trl/issues/835

This PR depends on https://github.com/huggingface/transformers/pull/27020 - with that PR, we introduce a new argument in the `gradient_checkpointing_enable()` API in order for users to pass `gradient_checkpointing_kwargs`. To fix some issues with DDP and gradient_checkpointing, it is recommended to use `use_reentrant=True` which is the fix for  https://github.com/huggingface/trl/issues/835 

Therefore I propose to expose an optional argument `gradient_checkpointing_kwargs` in `prepare_model_for_kbit_training`

cc @BenjaminBossan @pacman100

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Is this ready for review?

**Comment by younesbelkada** [comment]
Almost, I'll update the PR in a bit with more description

**Comment by younesbelkada** [comment]
https://github.com/huggingface/transformers/pull/27020 being merged, this PR is ready for review!

File: src/peft/utils/other.py [hunk 1]
```diff
@@ -63,18 +63,26 @@ def starcoder_model_postprocess_past_key_value(past_key_values):
...
 
     Args:
-        model, (`transformers.PreTrainedModel`):
+        model (`transformers.PreTrainedModel`):
             The loaded model from `transformers`
+        use_gradient_checkpointing (`bool`, *optional*, defaults to `True`):
+            If True, use gradient checkpointing to save memory at the expense of slower backward pass.
+        gradient_checkpointing_kwargs (`dict`, *optional*, defaults to `None`):
+            Keyword arguments to pass to the gradient checkpointing function, e.g. `use_reentrant=True`. Note this is
+            only available in the latest transformers versions.
```
Comment (BenjaminBossan) [inline src/peft/utils/other.py:79]: It would be better to specify the exact transformers version, because "latest" is relative.

File: src/peft/utils/other.py [hunk 2]
```diff
@@ -63,18 +63,26 @@ def starcoder_model_postprocess_past_key_value(past_key_values):
...
         head to fp32
 
     Args:
-        model, (`transformers.PreTrainedModel`):
+        model (`transformers.PreTrainedModel`):
             The loaded model from `transformers`
+        use_gradient_checkpointing (`bool`, *optional*, defaults to `True`):
+            If True, use gradient checkpointing to save memory at the expense of slower backward pass.
+        gradient_checkpointing_kwargs (`dict`, *optional*, defaults to `None`):
+            Keyword arguments to pass to the gradient checkpointing function, e.g. `use_reentrant=True`. Note this is
```
Comment (BenjaminBossan) [inline src/peft/utils/other.py:78]: Could you explain what `use_reentrant` does?

File: src/peft/utils/other.py [hunk 3]
```diff
@@ -63,18 +63,26 @@ def starcoder_model_postprocess_past_key_value(past_key_values):
     return tuple(result)
 
 
-def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
+def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True, gradient_checkpointing_kwargs=None):
```
Comment (BenjaminBossan) [inline src/peft/utils/other.py:66]: I just now notice that we should probably mention that this function only works on transformers (-compatible) models, passing a custom model won't work here.

File: src/peft/utils/other.py [hunk 4]
```diff
@@ -86,19 +94,29 @@ def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
...
 
-        # enable gradient checkpointing for memory efficiency
-        model.gradient_checkpointing_enable()
+        # To support older transformers versions, check if the model supports gradient_checkpointing_kwargs
+        _supports_gc_kwargs = "gradient_checkpointing_kwargs" in list(
+            inspect.signature(model.gradient_checkpointing_enable).parameters
+        )
+
+        gc_enable_kwargs = (
+            {} if not _supports_gc_kwargs else {"gradient_checkpointing_kwargs": gradient_checkpointing_kwargs}
```
Comment (BenjaminBossan) [inline src/peft/utils/other.py:115]: If the user passes `gradient_checkpointing_kwargs` but they are not supported, they would just be silently ignored, right? I wonder if we should not raise an error instead and tell the user that to use that feature, they have to use the new transformers version?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for making the big fix in transformers and updating PEFT to enable it. I have only a few comments, please check them out.

File: src/peft/utils/other.py (hunk 3 above)
Comment (younesbelkada) [reply src/peft/utils/other.py:66]: Yes makes sense, updated the docstring accordingly

File: src/peft/utils/other.py (hunk 4 above)
Comment (younesbelkada) [reply src/peft/utils/other.py:118]: Makes sense, I think however that a warning is sufficient here telling users that the arg will be silently ignored

File: src/peft/utils/other.py (hunk 2 above)
Comment (younesbelkada) [reply src/peft/utils/other.py]: I referred users to check the pytorch documentation, let me know if you want me to detail more

**Review by BenjaminBossan — APPROVED** [review]
Thanks for adjusting the code. It now looks good to me.

**Comment by younesbelkada** [comment]
Thanks, I'll merge after finishing some experiments with https://github.com/huggingface/transformers/pull/27068 and https://github.com/huggingface/trl/pull/912

**Comment by pacman100** [comment]
> To fix some issues with DDP and gradient_checkpointing, it is recommended to use use_reentrant=True which is the fix for huggingface/trl#835

Hello Younes, thank you for fixing the gradient checkpointing related issues as per our discussions. 

Just a nit, here you meant `use_reentrant=False`, right?
